### PR TITLE
Fix locate command in presence of let punning

### PR DESCRIPTION
**Pull request by liam923**
Currently, locating the `x` on line 2 will return `Already at definition` in the below program:
```
1. let x = Some 1 in
2. let* x in ...
```
Since the user is already at the definition point, if they request jumping to the definition, they likely want to go to line 1. This PR changes the behavior to instead jump to the definition on line 1.

The testsuite is not set up to easily support testing with `ppx_let`, but I have created tests in our internal codebase for `ppx_let`.

### CODE REVIEW

**Comment by liam923** [comment]
The tests are passing for me locally but failing on CI for some reason. I'll look into this later.

**Comment by liam923** [comment]
This is ready for review now.

File: tests/test-dirs/let-punning.t/run.t [hunk 1]
```diff
@@ -0,0 +1,124 @@
+Test Merlin's behavior in the presense of let-punning
```
[the remark this answers is not in the document]
Comment (liam923) [reply tests/test-dirs/let-punning.t/run.t:1]: Yep thanks for making those improvements.

File: src/ocaml/preprocess/parser_raw.mly [hunk 2]
```diff
@@ -280,7 +280,7 @@ let mkexp_type_constraint ?(ghost=false) ~loc ~modes e t =
      let mk = if ghost then ghexp_with_modes else mkexp_with_modes in
      mk ~loc ~exp:e ~cty:(Some t) ~modes
   | Pcoerce(t1, t2)  ->
-     let mk = if ghost then ghexp else mkexp ?attrs:None in
+     let mk = if ghost then ghexp ?attrs:None else mkexp ?attrs:None in
```
[the remark this answers is not in the document]
Comment (liam923) [reply src/ocaml/preprocess/parser_raw.mly]: Good call

File: src/kernel/mbrowse.ml [hunk 3]
```diff
@@ -137,23 +142,41 @@ let compare_locations pos l1 l2 =
   | _, _ ->
       Lexing.compare_pos l2.Location.loc_end l1.Location.loc_end
 
-let best_node pos = function
+let compare_nodes ?(let_pun_behavior = Let_pun_behavior.Prefer_pattern) pos (n1, loc1) (n2, loc2) =
+  let is_ghost node (loc : Location.t) =
```
[the remark this answers is not in the document]
Comment (liam923) [reply src/kernel/mbrowse.ml]: I tried to clean this up a bit.

**Review by ncik-roberts — APPROVED** [review]
Thanks for looking into this.

**Pull request merged**